File: backend/app/routers/gallery.py
"""
Gallery router — public endpoints with tag filtering.
"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional

from app.database import get_db
from app.models.gallery import GalleryImage
from app.schemas.gallery import GalleryImageOut

router = APIRouter(prefix="/gallery", tags=["Gallery"])


@router.get("", response_model=list[GalleryImageOut])
async def list_gallery(
    tag: Optional[str] = Query(None, description="Filter by tag"),
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    """Get gallery images, optionally filtered by tag."""
    query = select(GalleryImage).where(GalleryImage.is_published == True)

    if tag and tag.lower() != "all":
        query = query.where(GalleryImage.tag == tag.lower())

    query = query.order_by(desc(GalleryImage.uploaded_at)).limit(limit)
    result = await db.execute(query)
    images = result.scalars().all()

    return [GalleryImageOut.model_validate(img) for img in images]


@router.get("/tags")
async def get_gallery_tags():
    """Get available gallery tag categories."""
    return {
        "tags": [
            {"value": "all", "label": "All"},
            {"value": "international", "label": "International"},
            {"value": "intuc", "label": "INTUC"},
            {"value": "union", "label": "Unions"},
            {"value": "press", "label": "Press Release"},
            {"value": "timeline", "label": "Timeline"},
            {"value": "others", "label": "Others"},
        ]
    }

File: backend-php/api/routes/gallery.php
<?php
/**
 * routes/gallery.php — Public gallery endpoints.
 *
 * GET /gallery       — list images (filter by tag)
 * GET /gallery/tags  — available tag list
 */

$db = get_db();

// GET /gallery/tags  — must come before /gallery catch
if ($method === 'GET' && $path === '/gallery/tags') {
    json_success([
        'tags' => [
            ['value' => 'all',           'label' => 'All'],
            ['value' => 'international', 'label' => 'International'],
            ['value' => 'intuc',         'label' => 'INTUC'],
            ['value' => 'union',         'label' => 'Unions'],
            ['value' => 'press',         'label' => 'Press Release'],
            ['value' => 'timeline',      'label' => 'Timeline'],
            ['value' => 'others',        'label' => 'Others'],
        ]
    ]);
}

// GET /gallery
if ($method === 'GET' && $path === '/gallery') {
    $tag   = $_GET['tag'] ?? null;
    $limit = min(max(1, (int) ($_GET['limit'] ?? 50)), 200);

    if ($tag && strtolower($tag) !== 'all') {
        $stmt = $db->prepare(
            'SELECT id, filename, original_name, tag, caption, uploaded_at, source_timeline_id
             FROM gallery_images
             WHERE is_published = 1 AND tag = ?
             ORDER BY uploaded_at DESC
             LIMIT ?'
        );
        $stmt->execute([strtolower($tag), $limit]);
    } else {
        $stmt = $db->prepare(
            'SELECT id, filename, original_name, tag, caption, uploaded_at, source_timeline_id
             FROM gallery_images
             WHERE is_published = 1
             ORDER BY uploaded_at DESC
             LIMIT ?'
        );
        $stmt->execute([$limit]);
    }

    json_success($stmt->fetchAll());
}

json_error("Not found: [$method] $path", 404);
